### ISSUE TITLE
`armhf` builds from new Docker container images need testing

### ISSUE THREAD

**Issue by zultron** [machinekit/machinekit-hal#163]
The CI system's current Docker images are getting old, but can't be updated now that the Emdebian project is defunct; see Dovetail-Automata/mk-cross-builder#3 for details.  I've updated the [mk-cross-builder repo][1] to use the Linaro toolchain for cross-building `armhf` packages for Jessie, and the `armhf_8` tag builds packages perfectly with the changes from machinekit/machinekit#1425; see the [test images][2] in my personal Docker hub.

Before we can bless the new images, however, build results of esp. the `armhf_8` image need to be tested, and I don't have any BeagleBone hardware to do it myself.  Once these can be verified, we can feel safer about switching the CI system over to the new images.  If someone with a BeagleBone could run some tests, it would be a big help.

Here are brief instructions:
- Check out the latest Machinekit `master` branch on an `amd64` PC with Docker installed
- Build `armhf` packages for Jessie using the test images
  - `./scripts/build_docker -i zultron/mk-cross-builder -t armhf_8 -c deb`
  - Packages will be built into the parent directory
- Load the Beaglebone Debian Jessie [Machinekit image][4] onto a MicroSD card if you don't have one already, and start up the Beaglebone
- Copy the packages you just built over to the Beaglebone in your favorite way, and install the Xenomai flavor
  - `sudo dpkg -i machinekit{,-xenomai}_*~jessie_armhf.deb`
- Test Machinekit in your favorite way
  - Run a sim config
  - Run `latency-test`
  - Copy your real hardware config over and run it

Thanks!

[1]: https://github.com/zultron/mk-cross-builder/
[2]: https://hub.docker.com/r/zultron/mk-cross-builder/tags/
[3]: https://github.com/machinekit/machinekit/pull/1425/commits/4a83b9ef
[4]: https://elinux.org/Beagleboard:BeagleBoneBlack_Debian#BBW.2FBBB_.28All_Revs.29_Machinekit

**Comment by ArcEye** [comment]
I tested the build using your new dockers a couple of days ago and all succeeded.

Unfortunately I don't have the hardware to test the packages as per above.

**Comment by luminize** [comment]
I have a BBB, I’ll try to find some time to test as per instructions. Pretty busy right now so if somebody beats me that’s ok with me.

**Comment by ArcEye** [comment]
Thanks @luminize 

To make it easier for yourself, or anyone else testing, I have built the packages and they can be downloaded from here http://mgware.co.uk/armhf-jessie

**Comment by dkhughes** [comment]
I'm back from a trip. I will test the images out on the zynq platforms and report back. I can also update the zynq kernel / uboot builders to use gcc4 instead of gcc5.3 like they do currently for consistency in an image.

**Comment by zultron** [comment]
Thanks, @ArcEye, @luminize, @dkhughes for helping out with this.

@luminize Could you do me one extra favor that takes advantage of your Docker experience and BB hardware?  I have a patch for an upcoming PR that fixes `make install` of the `prubin` files, and I want to make sure I got the symlinks right.  If you wouldn't mind pulling the [`robots` branch][1] from my repo and building packages as above, I'd really appreciate it.  (And if it's too much hassle, that's ok; you're totally off the hook!)

[1]: https://github.com/zultron/machinekit/tree/robots

**Comment by luminize** [comment]
no problem, I'll do that

**Comment by luminize** [comment]
**First: the jessie image.**
I followed the instructions from John (original first comment) and made the debs for the robot branch https://github.com/machinekit/machinekit-hal/issues/163#issuecomment-446052901 copied the debs over to the BBB.

Because there's already a machinekit binary, I got some small issues, but after `sudo apt-get -f install` 
the packages `libczmq4` and `yapps2-runtime` get installed and I'm able to start machinekit.

The axis config starts up, but after homing and pressing the run button I get some messages telling me "program exceeds machine maximum on axis Z".  Pressing "run anyway" gives me "Linear move on line 0 would exceed joint 2's positive limit" but that seems to me config related.

**Comment by luminize** [comment]
So I guess this is _only_ for Jessie images, in that case I'm done, or do I need to test for stretch images too @zultron ?

**Comment by dkhughes** [comment]
Yes, the issue was the `emdebian` repos were dropped, and the cross compilation tools were changed to allow the docker image to build without error for jessie.

I have the packages running on zynq as well, but I haven't been able to test as much as I would like since I hit some issues unrelated. I didn't see the package conflicts when installing over the upstream packages - when was your base image built?

**Comment by luminize** [comment]
>  - when was your base image built?

From memory: beginning of January. it gets built once a month iirc

**Comment by zultron** [comment]
@luminize, thanks for firing up the sim build on the BBB Jessie image.  The mk-cross-builder changes were pretty much limited to swapping out the cross-build tool chain for Jessie `armhf` architecture, and I'd expect failures to be immediate and obvious, e.g. issues with dynamic library paths or `ld.so` symbols that would prevent MK from even starting up.

Since you've verified there are none of these issues, I'll go ahead and push the changes, which will trigger the MK CI system's Docker images to be updated over at the [DA repo][1].  Just to be on the safe side, I'll make copies the old images in the unlikely case we find we need to revert.

[1]: https://hub.docker.com/r/dovetailautomata/mk-cross-builder/

**Comment by ArcEye** [comment]
> I'll make copies the old images in the unlikely case we find we need to revert.

I have copies too, which I use on my Jenkins CI server.  They are altered very slightly with the addition of a couple of tools, but essentially the same.  So we have some resilience.

**Comment by zultron** [comment]
I tossed off a PR over at Dovetail-Automata/mk-cross-builder#4.  :D  (I hope that repo can someday be transferred over to the MK organization and it be taken over by the community.)

As it says there, in case these changes need to be reverted, previous images are saved in the [original repo][1], tagged as `*_old`.  Along with @ArcEye's copies, we should be safe if these ever need reverting.

Once the images are done building on the Docker hub, they automatically go live, successively over the next couple of hours.  First thing I'll do is tickle the [cpusets PR][2] after that; assuming all goes well & the PR is merged, I'll close out this issue.

[1]: https://hub.docker.com/r/dovetailautomata/mk-cross-builder/tags
[2]: https://github.com/machinekit/machinekit/pull/1426

**Comment by ArcEye** [comment]
The new dockers caused me a problem, because my server has limited disk space.
The pull of the new ones caused space to run out mid run and choked out Jenkins :(

I have resolved, by deleting all the old ones and doing fresh pulls of the new ones and it is all building properly again.

Don't think we need worry about the Machinekit server, it has acres of disk space to spare and
the builds have all concluded successfully
https://jenkins.machinekit.io/job/machinekit-builder/

**Comment by zultron** [comment]
@ArcEye, thanks for taking care of the Jenkins problems.  Stale CI images will become a bigger headache if we rebuild them often, but hopefully the current trend holds and it'll be a year or two until the next time.

I'm closing out this issue.  If we discover problems with the images down the road, a new issue can be opened.  Thanks for the help, @luminize @ArcEye @dkhughes!
